File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.java
import java.util.Queue;
import java.util.LinkedList;

class Solution {
    public int solution(String[] board) {
        int answer = -1;
        int s_r = 0;
        int s_c = 0;
        int e_r = 0;
        int e_c = 0;
        int N = board.length;
        int M = board[0].length();
        
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (board[i].charAt(j) == 'R') {
                    s_r = i;
                    s_c = j;
                } else if (board[i].charAt(j) == 'G') {
                    e_r = i;
                    e_c = j;
                }
            }
        }
        
        int[] dr = new int[] {-1, 1, 0, 0};
        int[] dc = new int[] {0, 0, -1, 1};
        
        boolean[][] visited = new boolean[N][M];
        visited[s_r][s_c] = true;
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[] {s_r, s_c, 0});
        
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            if (cur[0] == e_r && cur[1] == e_c) {
                answer = cur[2];
                break;
            }
            
            for (int k = 0; k < 4; k++) {
                int nr = cur[0] + dr[k];
                int nc = cur[1] + dc[k];
                
                while (true) {
                    if (0 <= nr && nr < N && 0 <= nc && nc < M) {
                        if (board[nr].charAt(nc) != 'D') {
                            nr += dr[k];
                            nc += dc[k];
                        } else {
                            break;
                        }
                    } else {
                        break;
                    }
                }
                
                nr -= dr[k];
                nc -= dc[k];
                if (!visited[nr][nc]) {
                    visited[nr][nc] = true;
                    q.add(new int[] {nr, nc, cur[2]+1});
                }
                
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.py
from collections import deque
def solution(board):
    answer = -1
    N = len(board)
    M = len(board[0])
    for i in range(N):
        for j in range(M):
            if board[i][j] == "R":
                s_r = i
                s_c = j
            elif board[i][j] == "G":
                e_r = i
                e_c = j
                
    visited = [[False] * M for _ in range(N)]
    q = deque()
    q.append((s_r, s_c, 0))
    visited[s_r][s_c] = True
    
    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]
    
    while q:
        r, c, cnt = q.popleft()
        if (r == e_r and c == e_c):
            answer = cnt
            break
        
        for k in range(4):
            nr = r + dr[k]
            nc = c + dc[k]
            while True:
                if 0 <= nr < N and 0 <= nc < M:
                    if board[nr][nc] != "D":
                        nr += dr[k]
                        nc += dc[k]
                        
                    else:
                        break
                else:
                    break
            nr -= dr[k]
            nc -= dc[k]
            if not visited[nr][nc]:
                q.append((nr, nc, cnt+1))
                visited[nr][nc] = True

    return answer